### PR TITLE
Add logging to spec

### PR DESCRIPTION
**Pull request by nahsra**
This PR adds the specification but it feels silly to prioritize perfection. It feels like we should roughly agree on the direction here, then revisit during implementation with small improvements where we find the spec is silent or unhelpful.

### CODE REVIEW

File: logging.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
+# Logging
+
+This specification dictates logging requirements for codemodder frameworks. Fameworks need to align on a general design, some important messages to log, and formats. This way, users get a relatively consistent experience, design problems are solved once and only once, and downstream tooling can make assumptions about expected output.
```
Comment (drdavella) [inline logging.md:3]:
```suggestion
This specification dictates logging requirements for codemodder frameworks. Frameworks need to align on a general design, some important messages to log, and output formats. This way, users get a relatively consistent experience, design problems are solved once and only once, and downstream tooling can make assumptions about expected output.
```

File: logging.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+## Output format
+
+The `--log-format` parameter tells the framework whether the output should be human readable by specifying `human` (which is the default) or `json`, which causes each message to be logged as its own self-contained JSON object.
```
Comment (drdavella) [inline logging.md:7]: Maybe make a reference to structured logging here?

File: logging.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+### `human` output
+
+For `human` output, we expect a message like the following:
+```
+$ ./run-my-codemod.sh (...) /my/project
+[starting] 
+codemodder: java/1.2
+```
+
+Notice there is very little context for the log message. The log message doesn't show the timestamp, origin, thread, or any other contextual metadata. We believe CLI users would prefer a very simple, undecorated set of messages.
```
Comment (drdavella) [inline logging.md:18]: Probably important to call out the absence of explicit "level" here too.

File: logging.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+ - `INFO` should contain information about changes
+ - `DEBUG` should contain diagnostic or forensic information
+
+## Verbosity
+
+The `--verbose` flag tells the codemodder framework to emit all log messages, regardless of their level. This flag has nothing to do with the output format, and strictly controls the volume of messages emitted.
+
+## Output streams
+
+Generally, logs are expected to go to `stdout`. When the user specificies that the output format should be `human`, or the default case where no output format is specified, it's preferred that error messages be emitted to `stderr`.
```
Comment (drdavella) [inline logging.md:41]:
```suggestion
Generally, logs are expected to go to `stdout`. When the user specificies that the output format should be `human`, including the default case where no output format is specified, it's preferred that error messages be emitted to `stderr`.
```

File: logging.md (hunk 3 above)
Comment (drdavella) [inline logging.md:18]: This feels like a slightly better wording to me but I'm not sure
```suggestion
Notice there is very little boilerplate for the log message. The log message doesn't show the timestamp, origin, thread, or any other contextual metadata. We believe CLI users would prefer a very simple, undecorated set of messages.
```

File: logging.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+running: pixee:javascript/limit-readline, pixee:javascript/secure-random
+including paths: **/*.js
+excluding paths: **/tests/**,**/test.js
+
+[scanning]
+codemod: pixee:javascript/limit-readline
+scanned: 100 files...
+scanned: 200 files...
+scanned: 300 files...
+failed:
```
Comment (drdavella) [inline logging.md:84]: I know this is just for the sake of example and it makes sense in that context, but in "real life" I would also expect to emit some error messages here related to the failures. I'm not sure how to call that out in the document though. I do think we need some language somewhere that says that codemodders are allowed to emit additional messages at various levels where appropriate (i.e. the spec doesn't strictly constrain the number/type of messages that can possibly be emitted).

**Review by drdavella — CHANGES_REQUESTED** [review]
This looks very good to me overall. I have some nitpicky comments about wording but also a few more substantive comments where we might want to add a little more detail/clarification to the spec.

File: logging.md (hunk 5 above)
Comment (nahsra) [reply logging.md:84]: Doesn't this bit communicate this?

> Frameworks can emit more messages, but the following messages serve as a reasonable first milestone or baseline for good-enough logging to support users and downstream tooling.

**Comment by nahsra** [comment]
> One other high-level comment which occurred to me recently and which we might want to include in the spec. The Python codemodder currently emits diffs to stdout which I find really helpful as a developer/user. Should we include that in the spec somehow? Should that occur only at `--verbose` and `--log-format=human`? Purely optional but with those constraints?

I like this. I added some strawman version. I think it should be present for verbose.

File: logging.md (hunk 5 above)
Comment (drdavella) [reply logging.md:84]: Yeah you're right, my bad.

**Review by drdavella — APPROVED** [review]
LGTM. I don't see any language about printing diffs in this version (not sure whether it got lost) but I'll approve and you can feel free to add it before merging if that seems appropriate.

**Comment by nahsra** [comment]
> LGTM. I don't see any language about printing diffs in this version (not sure whether it got lost) but I'll approve and you can feel free to add it before merging if that seems appropriate.

Whoops, never pushed the diff. It's on right now for verbose.

**Comment by drdavella** [comment]
Yup that looks good. Don't let me stop you from merging but do we want to make it officially part of the spec? Like `--verbose` + `--log-format=human` implies diffs will be logged? We can always go back and add this later if you just want to move forward now.

**Pull request merged**